module/icp/monocypher.c

```diff
@@ -0,0 +1,385 @@
+//    documentation and/or other materials provided with the
+//    distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "monocypher.h"
+
+/////////////////
+/// Utilities ///
+/////////////////
+#define FOR_T(type, i, start, end) for (type i = (start); i < (end); i++)
+#define FOR(i, start, end)         FOR_T(size_t, i, start, end)
+#define COPY(dst, src, size)       FOR(i__, 0, size) (dst)[i__] = (src)[i__]
+#define ZERO(buf, size)            FOR(i__, 0, size) (buf)[i__] = 0
+#define WIPE_CTX(ctx)              crypto_wipe(ctx   , sizeof(*(ctx)))
+#define WIPE_BUFFER(buffer)        crypto_wipe(buffer, sizeof(buffer))
+
+#ifndef MIN
+#define MIN(a, b)                  ((a) <= (b) ? (a) : (b))
+#endif
+#ifndef MAX
+#define MAX(a, b)                  ((a) >= (b) ? (a) : (b))
+#endif
+
+typedef int8_t   i8;
+typedef uint8_t  u8;
+typedef int16_t  i16;
+typedef uint32_t u32;
+typedef int32_t  i32;
+typedef int64_t  i64;
+typedef uint64_t u64;
+
+static const u8 zero[128] = {0};
+
+// returns the smallest positive integer y such that
+// (x + y) % pow_2  == 0
+// Basically, it's how many bytes we need to add to "align" x.
+// Only works when pow_2 is a power of 2.
+// Note: we use ~x+1 instead of -x to avoid compiler warnings
+static size_t align(size_t x, size_t pow_2)
+{
+    return (~x + 1) & (pow_2 - 1);
+}
+
+static u32 load32_le(const u8 s[4])
+{
+    return (u32)s[0]
+        | ((u32)s[1] <<  8)
+        | ((u32)s[2] << 16)
+        | ((u32)s[3] << 24);
+}
+
+static u64 load64_le(const u8 s[8])
+{
+    return load32_le(s) | ((u64)load32_le(s+4) << 32);
+}
+
+static void store32_le(u8 out[4], u32 in)
+{
+    out[0] =  in        & 0xff;
+    out[1] = (in >>  8) & 0xff;
+    out[2] = (in >> 16) & 0xff;
+    out[3] = (in >> 24) & 0xff;
+}
+
+static void load32_le_buf (u32 *dst, const u8 *src, size_t size) {
+    FOR(i, 0, size) { dst[i] = load32_le(src + i*4); }
+}
+
+static u32 rotl32(u32 x, u32 n) { return (x << n) ^ (x >> (32 - n)); }
+
+static int neq0(u64 diff)
+{   // constant time comparison to zero
+    // return diff != 0 ? -1 : 0
+    u64 half = (diff >> 32) | ((u32)diff);
+    return (1 & ((half - 1) >> 32)) - 1;
```

Comment:
More unnecessary function overhead.

Reply:
Also, why not just do `-(diff != 0)`?

Reply:
Constant time comparisons are used rarely enough that you never care about the overhead in practice. And I checked the generated code, it's all inlined anyway, and the compiler just do direct 64-bit loads and XORs them appropriately.

The reason I avoid the comparison operator is because compilers are more likely to replace them by an actual branch, which is potentially disastrous. Granted it's only one last branch at the very end of the comparison (regardless of whether it's all equal or different), so we _should_ be okay. But the pure arithmetic doesn't trigger Valgrind on uninitialised data (only branches & indices do), which make the ctgrind analysis less noisy (fewer false positives about variable-time code).

Reply:
> Constant time comparisons are used rarely enough that you never care about the overhead in practice. And I checked the generated code, it's all inlined anyway, and the compiler just do direct 64-bit loads and XORs them appropriately.

That is fair.

> The reason I avoid the comparison operator is because compilers are more likely to replace them by an actual branch, which is potentially disastrous.

I had not considered that. I just learned something new. Thanks.

> Granted it's only one last branch at the very end of the comparison (regardless of whether it's all equal or different), so we _should_ be okay. But the pure arithmetic doesn't trigger Valgrind on uninitialised data (only branches & indices do), which make the ctgrind analysis less noisy (fewer false positives about variable-time code).

This is also new to me. Thanks for writing this explanation.